### PR TITLE
feature: add semantic version types

### PR DESCRIPTION
**Pull request by mnoman09**
## Summary
- add greater than or equal and less than or equal for numbers
- add semantic version match types.

### CODE REVIEW

File: OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs [hunk 1]
```diff
@@ -302,6 +309,38 @@ public void TestGTMatcherReturnsTrueWhenAttributeValueIsGreaterThanConditionValu
         }
 
         #endregion // GTMatcher Tests
+        
+        #region GEMatcher Tests
+
+        [Test]
+        public void TestGEMatcherReturnsFalseWhenAttributeValueIsLessButTrueForEqualToConditionValue()
+        {
+            Assert.That(GECondition.Evaluate(null, new UserAttributes { { "distance_ge", 5 } }, Logger), Is.False);
```
Comment (msohailhussain) [inline OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs:318]: use `Assert.IsTrue or Assert.IsFalse`

File: OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs [hunk 2]
```diff
@@ -30,12 +30,19 @@ public class ConditionEvaluationTest
...
+        private BaseCondition LECondition = new BaseCondition { Name = "distance_le", Value = 10, Match = "le", Type = "custom_attribute" };
         private BaseCondition ExactStrCondition = new BaseCondition { Name = "browser_type", Value = "firefox", Match = "exact", Type = "custom_attribute" };
         private BaseCondition ExactBoolCondition = new BaseCondition { Name = "is_registered_user", Value = false, Match = "exact", Type = "custom_attribute" };
         private BaseCondition ExactDecimalCondition = new BaseCondition { Name = "pi_value", Value = 3.14, Match = "exact", Type = "custom_attribute" };
         private BaseCondition ExactIntCondition = new BaseCondition { Name = "lasers_count", Value = 9000, Match = "exact", Type = "custom_attribute" };
         private BaseCondition InfinityIntCondition = new BaseCondition { Name = "max_num_value", Value = 9223372036854775807, Match = "exact", Type = "custom_attribute" };
+        private BaseCondition SemVerLTCondition = new BaseCondition { Name = "semversion_lt", Value = "3.7.1", Match = "semver_lt", Type = "custom_attribute" };
```
Comment (msohailhussain) [inline OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs:41]: Can you please append values, so unit test will have more clarity.

File: OptimizelySDK/AudienceConditions/BaseCondition.cs [hunk 3]
```diff
@@ -156,6 +170,35 @@ public class BaseCondition : ICondition
             return Convert.ToDouble(attributeValue) > Convert.ToDouble(Value);
         }
 
+        public bool? GreaterOrEqualThanEvaluator(object attributeValue, ILogger logger)
```
Comment (msohailhussain) [inline OptimizelySDK/AudienceConditions/BaseCondition.cs:173]: Am not in favor of all these evaluator or we should put common logic in a separate method, the only evaluation logic should be kept inside every evaluator.

File: OptimizelySDK/AudienceConditions/SemanticVersion.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Collections.Generic;
+
+namespace OptimizelySDK.AudienceConditions
+{
+    public static class SemanticVersionExtension
```
Comment (msohailhussain) [inline OptimizelySDK/AudienceConditions/SemanticVersion.cs:21]: add Documentation class.

File: OptimizelySDK/AudienceConditions/SemanticVersion.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    {
+        public const char BuildSeparator = '+';
+        public const char PreReleaseSeparator = '-';
+
+        public static bool ContainsWhiteSpace(this string semanticVersion)
```
Comment (msohailhussain) [inline OptimizelySDK/AudienceConditions/SemanticVersion.cs:26]: Document all methods.

File: OptimizelySDK/AudienceConditions/SemanticVersion.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            if (!string.IsNullOrEmpty(versionSuffix))
+            {
+                versionParts.Add(versionSuffix);
+            }
+
+            return versionParts.ToArray();
+        }
+    }
+
+    public class SemanticVersion : IComparable<SemanticVersion>
```
Comment (msohailhussain) [inline OptimizelySDK/AudienceConditions/SemanticVersion.cs:89]: Documentation missing.

File: OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs (hunk 1 above)
Comment (mnoman09) [reply OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs:318]: Evaluate return bool? and Assert.IsTrue expects bool. That is the reason I used Assert.That

File: OptimizelySDK/AudienceConditions/BaseCondition.cs (hunk 3 above)
Comment (msohailhussain) [reply OptimizelySDK/AudienceConditions/BaseCondition.cs:154]: Lets talk offline. CommonEvaluator should return either 0, 1, -1 or raise exception. We shouldn't parse anything in GE, LE, EQ (SemVer).

File: OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs [hunk 7]
```diff
@@ -334,5 +373,230 @@ public void TestLTMatcherReturnsTrueWhenAttributeValueIsLessThanConditionValue()
...
+
+        [Test]
+        public void TestLEMatcherReturnsTrueWhenAttributeValueIsLessThanConditionValue()
+        {
+            Assert.IsTrue(LECondition.Evaluate(null, new UserAttributes { { "distance_le", 5 } }, Logger) ?? false);
+        }
+
+        #endregion // LEMatcher Tests
+
+        #region SemVerLTMatcher Tests
```
Comment (thomaszurkan-optimizely) [inline OptimizelySDK.Tests/AudienceConditionsTests/ConditionEvaluationTest.cs:408]: Can we add some invalid attribute tests? (i.e.  "3.0-", 3.x", 3.2, "3.0.0+", "+")?

File: OptimizelySDK/AudienceConditions/SemanticVersion.cs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                var partialVersionParts = version.Split(version.IsPreRelease() ?
+                    PreReleaseSeparator : BuildSeparator);
+                if (partialVersionParts.Length <= 1)
+                {
+                    // throw error
+                    throw new Exception("Invalid Semantic Version.");
+                }
+                // major.minor.patch
+                var versionPrefix = partialVersionParts[0];
+                versionSuffix = partialVersionParts[1];
```
[the remark this answers is not in the document]
Comment (msohailhussain) [reply OptimizelySDK/AudienceConditions/SemanticVersion.cs:87]: good catch Tom.

**Pull request merged**